### PR TITLE
[IMP] module installation time due to a recursive strategy

### PR DESCRIPTION
**Pull request by etobella**
Description of the issue/feature this PR addresses:
This PR reduces the time of installation of a new module.

Current behavior before PR:
If you have a module with several dependencies or a complex dependency graph, the system repeats the number of checks of a module. For example: we have 4 modules A, B, C and D, where A depends on B and C, B depends on D and C depends on D. Then, with the current code, we will check once if A is installed, once if B is installed, once if C is installed and twice id D is installed.

Desired behavior after PR is merged:
We only check once if a module is installed.

I have signed the CLA, but there is some issues with robodoo #29593

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by etobella** [comment]
@moylop260 You will love this one, it took more than 10 minutes to start the install of a system with more than 170 installed modules, now it takes like 5 seconds.
It was testing a lot of times if base was installed.

**Comment by Yenthe666** [comment]
@mart-e can you please give this one some attention when you have some time? I like the idea.

**Comment by etobella** [comment]
Well, you can see the change of behaviour on travis:
https://travis-ci.com/etobella/odoo-slow-install2/builds/95822137
With my patch it takes 2 seconds to restart the server in order to reinstall:
https://travis-ci.com/etobella/odoo-slow-install2/jobs/167271421#L1032
With the current system it never loads again:
https://travis-ci.com/etobella/odoo-slow-install2/jobs/167271420#L1018

**Comment by mart-e** [comment]
Hi,

Thank you for the PR.

> You will love this one, it took more than 10 minutes to start the install of a system with more than 170 installed modules, now it takes like 5 seconds.

Could you elaborate that part? While your patch does reduce the number of call to `_state_update`, it does not reduce the number of write operations made (`if module.state in states_to_update`).
How do you get a 10min to start the install? Oo

**Comment by etobella** [comment]
The problem I found was that in order to check if a module is installed, it must check all the dependencies (and the dependencies of the dependencies and so on). But the process does not remember which modules where checked, so it was checking the same module several times.
I may try to show it on an example:
we have 4 modules. a depends on b and c, b depends on base and c depends on base. When we try to install a it will do the following:
- a is installed? No
- check the dependencies of a [b, c]
  - b is installed? No
  - check the dependencies of b [base]
    - base is installed? Yes
    - check the dependencies of base []
  - c is installed? No
  - check the dependencies of c [base]
    - base is installed? Yes
    - check the dependencies of base []

As you can see, base was checked twice. If the tree is more complex it will be checked much more times...
On the repositories https://github.com/etobella/odoo-slow-install and https://github.com/etobella/odoo-slow-install2 I was creating this kind of modules and showing the issue.

**Comment by mart-e** [comment]
I do agree that it is more simple and indeed could be a nice improvement of the complexity of the algorithm but I was wondering if the patch had a noticeable impact.

I made a quick try with your modules, while several unnecessary verifications were made, it was still quite fast.
Do you really get 10min with some modules? That would be problematic.

**Comment by etobella** [comment]
As you can check on the previous posts you can check the travis without the patch:
https://travis-ci.com/etobella/odoo-slow-install2/jobs/167271420#L1018 (it stops because it has been 10 minutes waiting)
and with the patch:
https://travis-ci.com/etobella/odoo-slow-install2/jobs/167271421#L1032

**Comment by pedrobaeza** [comment]
@etobella the problem is not there in dependency check. It should be due to other things in your pipeline that gets to a deadlock (maybe the know problem when printing PDFs).

**Comment by etobella** [comment]
@pedrobaeza My first thought was that it was related to something I made, for that reason I created the odoo-slow-install repositories. The modules there do nothing, as they are only creating dependencies and the issue can be reproduced there too. Also, once I applied the patch it solved the issue in both systems (my system and the odoo-slow-install repo). 

Also, you can debug the system and you will see that it is checking multiple times if a module is installed. So, IMHO, there is a problem there and it can be related to my issue.

**Comment by pedrobaeza** [comment]
One thing is sure: dependencies check is not taking 10 minutes.

**Comment by etobella** [comment]
@pedrobaeza Can you check the following log please?
https://travis-ci.com/etobella/odoo-slow-install2/jobs/169023032
The only difference is that logging is writing every check of a module.

**Comment by pedrobaeza** [comment]
It seems the travis log is not the correct one for pointing that.

**Comment by etobella** [comment]
Sorry: https://api.travis-ci.com/v3/job/169023032/log.txt

**Comment by pedrobaeza** [comment]
Please look to timestamp. There's only 1 minute last doing the check, not 10 minutes, and it's slower due to the verbose logging. Any way, I'm not against this improvement, but there's more things in your Travis stack.

**Comment by etobella** [comment]
There is only a minute because travis broke. You can see it on the end:
`The job exceeded the maximum log length, and has been terminated.`
And you can see the proposed change. Also, it had to install 'module_a11', 'module_a12', 'module_b11', 'module_b12', 'module_c11', 'module_c12',

As you can see, 'module_a12', 'module_b11', 'module_b12', 'module_c11', 'module_c12', have not been checked... So It should take, at least,1 min 10 seconds to check only module_a11 (it did not start to check any other module, so you can see it should take more than that for one of the modules, as it never started the second one)

**Comment by moylop260** [comment]
FYII have reproduced the issue from a travis2docker environment using:
`t2d https://github.com/etobella/odoo-slow-install2 master`

But you can reproduce it locally using the following repositories:
```bash
# Clone tree of modules with case of use
git clone https://github.com/etobella/odoo-slow-install2 /tmp/odoo-slow-install2
git clone https://github.com/etobella/odoo-slow-install /tmp/odoo-slow-install

# Start a database template installing main modules
~/odoo_git/odoo-bin --addons-path=/tmp/odoo-slow-install2,/tmp/odoo-slow-install,~/odoo_git/addons -d slow_install -i module_a1,module_a6,module_c2,web_planner,module_b2,module_c9,account,product,module_b1,module_c1,module_b10,module_c3,module_a8,base_setup,mail,web_tour,http_routing,module_a7,module_a3,module_b9,module_c10,module_b6,decimal_precision,module_a10,module_c4,analytic,module_b5,module_a5,module_c,module_c5,portal,module_c6,module_b,module_a,module_c8,base,module_b7,module_a4,module_a2,module_b3,module_a9,web,bus,module_b4,module_c7,module_b8 --stop-after-init

# Start a database installing work modules
~/odoo_git/odoo-bin --addons-path=/tmp/odoo-slow-install2,/tmp/odoo-slow-install,~/odoo_git/addons -d slow_install -i module_a11,module_a12,module_b11,module_b12,module_c11,module_c12 --stop-after-init
```

You will see the following log:
`2019-01-09 18:59:16,507 2383 INFO openerp_test odoo.addons.base.module.module: ALLOW access to module.button_install on ['module_a11', 'module_a12', 'module_b11', 'module_b12', 'module_c11', 'module_c12'] to user admin #1 via n/a`
A lot of time without this patch.

File: odoo/addons/base/module/module.py [hunk 1]
```diff
@@ -348,35 +348,36 @@ def check_external_dependencies(cls, module_name, newstate='to install'):
...
         if level < 1:
             raise UserError(_('Recursion error in modules dependencies !'))
 
         # whether some modules are installed with demo data
         demo = False
-
-        for module in self:
+        if not checked:
+            checked = self.browse()
+        for module in self.filtered(lambda r: r not in checked):
```
Comment (moylop260) [inline odoo/addons/base/module/module.py]: What about `for module in self - checked`?

**Comment by pedrobaeza** [comment]
I don't think it's a question of efficiency of code, but a deadlock provoked by recursive calls that this refactoring avoids. @moylop260 have you checked this in your test?

**Comment by moylop260** [comment]
@pedrobaeza 
Yes, you are right.

This patch avoids an infinite loop by recursive calls.
The travis stop after 10 minutes, but if you wait to finish more time. It won't finish.

**Comment by etobella** [comment]
@mart-e 
Have you checked the deadlock?

**Comment by etobella** [comment]
Ping @Yajo

**Comment by mart-e** [comment]
Hi,

Better late than never. Thanks for the patch and the information on how to reproduce.
I have cherry picked it and merged into master at https://github.com/odoo/odoo/commit/2486259ef99ec85cf1c0a8200e31d350c582fd4c

**Comment by Elkasitu** [comment]
Hi,

This change seems to break the installation of demo data in certain cases, e.g.:

`./odoo-bin ... -i website --stop-after-init` -- website demo installed correctly
`./odoo-bin ... -i website_blog --stop-after-init` -- website and website_blog demo data installed correctly
`./odoo-bin ... -i website --stop-after-init && ./odoo-bin ... -i website_blog --stop-after-init` -- website demo installed correctly but website_blog demo not installed at all

I do not see a huge performance difference in Odoo standard either, @etobella can you elaborate on your use-case? are we definitely sure that your problem is not a case of a dependency cycle?

I'm afraid that unless there's a clear fix or benefit, this patch might be reverted

/cc @JKE-be @mart-e

**Comment by etobella** [comment]
The performabce issue can be seen with the logs attached. This solves a recursion issue. I will check why it is giving issues on your check, but I think it is unrelated...

**Comment by etobella** [comment]
@Elkasitu 
You are right, there is an issue, I will create a PR today in order to fix it. Sorry for the issue.

**Pull request closed without merging**